recon/src/recon/durability.rs

```diff
@@ -0,0 +1,280 @@
+        )";
+
+        self.conn
+            .execute(
+                CREATE_RECON_TABLE,
+                (), // empty list of parameters.
+            )
+            .unwrap();
+    }
+}
+
+impl<K, H> Store for SQLiteStore<K, H>
+where
+    K: Key + From<Vec<u8>>,
+    H: AssociativeHash,
+{
+    type Key = K;
+    type Hash = H;
+
+    // Ok(true): inserted the key
+    // Ok(false): did not insert the key ConstraintViolation
+    // Err(e): sql error
+    fn insert(&mut self, key: &Self::Key) -> anyhow::Result<bool> {
+        let hash = H::digest(key);
+
+        let resp = self.conn.execute(
+            "INSERT INTO recon (
+                sort_key, event_id,
+                ahash_0, ahash_1, ahash_2, ahash_3,
+                ahash_4, ahash_5, ahash_6, ahash_7,
+                block_retrieved
+            ) VALUES (
+                ?, ?,
+                ?, ?, ?, ?,
+                ?, ?, ?, ?,
+                ?
+            )",
+            (
+                &self.sort_key.as_bytes(),
+                key.as_bytes(),
+                hash.as_u32s()[0],
+                hash.as_u32s()[1],
+                hash.as_u32s()[2],
+                hash.as_u32s()[3],
+                hash.as_u32s()[4],
+                hash.as_u32s()[5],
+                hash.as_u32s()[6],
+                hash.as_u32s()[7],
+                false,
+            ),
+        );
+        match resp {
+            Ok(_) => Ok(true),
+            Err(e) => {
+                if e.sqlite_error_code() == Some(ErrorCode::ConstraintViolation) {
+                    Ok(false)
+                } else {
+                    Err(anyhow!("SQL Error: {}", e))
+                }
+            }
+        }
+    }
+
+    fn hash_range(&self, left_fencepost: &Self::Key, right_fencepost: &Self::Key) -> Self::Hash {
+        let query = "
+        SELECT
+            TOTAL(ahash_0) & 0xFFFFFFFF, TOTAL(ahash_1) & 0xFFFFFFFF,
+            TOTAL(ahash_2) & 0xFFFFFFFF, TOTAL(ahash_3) & 0xFFFFFFFF,
+            TOTAL(ahash_4) & 0xFFFFFFFF, TOTAL(ahash_5) & 0xFFFFFFFF,
+            TOTAL(ahash_6) & 0xFFFFFFFF, TOTAL(ahash_7) & 0xFFFFFFFF
+        FROM
+            recon
+        WHERE
+            sort_key = ? AND
+            event_id > ? AND event_id < ?; 
+        ";
+        let mut stmt = self.conn.prepare(query).unwrap();
+        let mut rows = stmt
+            .query((
+                self.sort_key.as_bytes(),
+                left_fencepost.as_bytes(),
+                right_fencepost.as_bytes(),
+            ))
+            .unwrap();
+        let row = rows.next().unwrap().unwrap();
+        let bytes: [u32; 8] = [
+            row.get(0).unwrap(),
+            row.get(1).unwrap(),
+            row.get(2).unwrap(),
+            row.get(3).unwrap(),
+            row.get(4).unwrap(),
+            row.get(5).unwrap(),
+            row.get(6).unwrap(),
+            row.get(7).unwrap(),
+        ];
+        H::from(bytes)
+    }
+
+    fn range(
```

Comment:
So rather than changing the type of range, one way to do this is we return a SqliteRange type that then implements `Iterator<Item = &Self::Key>`.  The store trait then has a RangeType with those bounds.

```
trait Store {
    type Range;
    
    fn range(...) where Self::Range: for<'a> Iterator<Item = &'a Self::Key>;
}

impl Store for SqliteStore {
   type Range = SqliteRange
}```

Reply:
Although I think with that, we'd probably want to make Range a trait, and then recon returns a Range, which then implements the functionality that Recon type has.